lib/cc/analyzer/path_filter.rb

```diff
@@ -0,0 +1,65 @@
+module CC
+  module Analyzer
+    class PathFilter
+      attr_reader :paths
+
+      def initialize(paths)
+        @paths = paths
+      end
+
+      def raise_if_any_unreadable_files
+        paths.each do |path|
+          if !File.directory?(path) && !FileUtils.readable_by_all?(path)
+            raise CC::Analyzer::UnreadableFileError, "Can't read #{path}"
+          end
+        end
+
+        self
+      end
+
+      def reject_unreadable_paths
+        @paths = paths.reject { |path| unreadable_path_entries.include?(path) }
+        self
+      end
+
+      def reject_paths(ignore_paths)
+        @paths = paths.reject { |path| ignore_paths.include?(path) }
+        self
+      end
+
+      def select_readable_files
+        @paths = paths.select { |path| FileUtils.readable_by_all?(path) }
+        self
+      end
+
+      def reject_symlinks
+        @paths = paths.reject { |path| File.symlink?(path) }
+        self
+      end
+
+      def reject_globs(globs)
+        @paths = paths.reject { |path| matches_globs?(path, globs) }
+        self
+      end
+
+      private
+
+      def unreadable_path_entries
+        @_unreadable_path_entries ||=
+          unreadable_paths.map { |path| PathEntries.new(path).entries }.flatten
+      end
+
+      def unreadable_paths
+        paths.select do |path|
+          File.directory?(path) && !FileUtils.readable_by_all?(path)
+        end
+      end
+
+      def matches_globs?(path, globs)
+        globs.any? do |exclude_path|
+          File.fnmatch(exclude_path, path)
```

Comment:
I think these are the `exclude_paths` that users give us.

I'll look into PathPatterns and see if that works here.